### ISSUE TITLE
Consider supporting newer TLS cipher suites

### ISSUE THREAD

**Issue by moderation** [lyft/envoy#705]
https://github.com/lyft/envoy/blob/master/source/common/ssl/context_config_impl.cc specifies supported cipher suites for Envoy.

The recent Go 1.8 release included support for the CHACHA20_POLY1305 cipher suites.

The following modern cipher suites are recommended by the [Mozilla SSL Configuration Generator](https://mozilla.github.io/server-side-tls/ssl-config-generator/)
```
ECDHE-ECDSA-AES128-GCM-SHA256
ECDHE-ECDSA-AES128-SHA256
ECDHE-ECDSA-AES256-GCM-SHA384
ECDHE-ECDSA-AES256-SHA384
ECDHE-ECDSA-CHACHA20-POLY1305
ECDHE-RSA-CHACHA20-POLY1305
```
BoringSSL support all of these ciphers.

**Comment by moderation** [comment]
Have done some testing and the following ChaCha ciphers result in a  runtime failure
```
ECDHE-ECDSA-CHACHA20-POLY1305
ECDHE-RSA-CHACHA20-POLY1305
```
Error
```
[2017-04-06 09:28:00.296][17553][critical][main] error initializing configuration 'sync-front-envoy.json': Unknown cipher specified in cipher suites ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-SHA256:ECDHE-RSA-AES128-SHA:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA:AES128-SHA
```
In theory BoringSSL [supports these ciphers](https://boringssl.googlesource.com/boringssl/+/2924/include/openssl/tls1.h#602). Will do further testing.

**Comment by PiotrSikora** [comment]
Thanks for bringing this up, @moderation. I'll look into why it doesn't work and update defaults.

@mattklein123 what are you thoughts on keeping OpenSSL support around? I'm asking, because if we're adding `CHACHA20-POLY1305`, then it would be useful to do that using "equal-preference groups", which allow clients to chose between CHACHA20 & AES-GCM (depending on hardware support for AES-NI, etc.).

However, OpenSSL doesn't support this yet (see: https://github.com/openssl/openssl/issues/541), so it would require another `#ifdef` for BoringSSL vs OpenSSL, which as far as I recall, you're not the biggest fan of.

Also, feel free to assign this to me.

**Comment by mattklein123** [comment]
@PiotrSikora I'm not sure how much people care about openssl. Personally, I feel like if we tell people what SHA of boringssl to use it's probably fine for us to drop openssl support. Let me poll a few people.

**Comment by mattklein123** [comment]
@PiotrSikora as part of this, can you also advise on how best to handle TLS 1.3? Should we make it opt in on the context for now?

**Comment by moderation** [comment]
@PiotrSikora interesting that you mention equal preference groups as I just saw AGL on Twitter [mention them](https://twitter.com/agl__/status/851462096185303040). I was going to retest the ChaCha Poly ciphers with BoringSSL master but I can't see why they fail with the current SHA.  My current working config is:
```
const std::string ContextConfigImpl::DEFAULT_CIPHER_SUITES = "ECDHE-ECDSA-AES256-GCM-SHA384:"
                                                             "ECDHE-ECDSA-AES256-SHA384:"
                                                             "ECDHE-ECDSA-AES128-GCM-SHA256:"
                                                             "ECDHE-ECDSA-AES128-SHA256:"
                                                             "ECDHE-RSA-AES128-GCM-SHA256:"
                                                             "ECDHE-RSA-AES128-SHA256:"
                                                             "ECDHE-RSA-AES128-SHA:"
                                                             "ECDHE-RSA-AES256-GCM-SHA384:"
                                                             "ECDHE-RSA-AES256-SHA384:"
                                                             "ECDHE-RSA-AES256-SHA:"
                                                             "AES256-GCM-SHA384:"
                                                             "AES256-SHA:"
                                                             "AES128-GCM-SHA256:"
                                                             "AES128-SHA256:"
                                                             "AES128-SHA";
```

**Comment by PiotrSikora** [comment]
See: https://github.com/lyft/envoy/pull/750, https://github.com/lyft/envoy/pull/751 & https://github.com/lyft/envoy/pull/752.

@moderation For some reason, I was unable to cc you on those pull requests.

**Comment by moderation** [comment]
Looks great @PiotrSikora. Do you think we need to add the additional curves as per my edited comment at the [top of this thread](https://github.com/lyft/envoy/issues/705#issue-219761154)?

**Comment by PiotrSikora** [comment]
They are all included in #752.
